app_flutter/lib/widgets/filter_property_sheet.dart

```diff
@@ -0,0 +1,330 @@
+  }
+
+  @override
+  void dispose() {
+    widget.propertySource.removeListener(_update);
+
+    super.dispose();
+  }
+
+  static const TextStyle _labelStyle = TextStyle(
+    color: Colors.black,
+    fontSize: 16.0,
+    fontWeight: FontWeight.normal,
+    fontStyle: FontStyle.normal,
+    decoration: TextDecoration.none,
+  );
+
+  void _update() {
+    setState(() {});
+  }
+
+  Widget _pad(Widget child, Alignment alignment) {
+    return Container(
+      padding: const EdgeInsets.all(5.0),
+      child: child,
+      alignment: alignment,
+    );
+  }
+
+  TableRow _makeRow(String label, Widget editable) {
+    return TableRow(
+      children: <Widget>[
+        _pad(Text(label, style: _labelStyle), Alignment.centerRight),
+        _pad(editable, Alignment.centerLeft),
+      ],
+    );
+  }
+
+  TableRow _makeTextFilterRow(RegExpFilterProperty property) {
+    return _makeRow(
+      property.label,
+      TextField(
+        autofocus: true,
+        controller: property.controller,
+        decoration: const InputDecoration(
+          hintText: '(JavaScript regular expression)',
+        ),
+        onChanged: (String value) => property.value = value,
+      ),
+    );
+  }
+
+  TableRow _makeBoolRow(BoolFilterProperty property) {
+    return _makeRow(
+      property.label,
+      Checkbox(
+        value: property.value,
+        onChanged: (bool newValue) => property.value = newValue,
+      ),
+    );
+  }
+
+  Widget _makeLoneCheckbox(BoolFilterProperty property) {
+    return Row(
+      mainAxisSize: MainAxisSize.min,
+      children: <Widget>[
+        Text(property.label, style: _labelStyle),
+        Checkbox(
+          value: property.value,
+          onChanged: (bool newValue) => property.value = newValue,
+        ),
+      ],
+    );
+  }
+
+  TableRow _makeTableRow(FilterPropertyNode property) {
+    if (property is RegExpFilterProperty) {
+      return _makeTextFilterRow(property);
+    }
+    if (property is BoolFilterProperty) {
+      return _makeBoolRow(property);
+    }
+    if (property is BoolFilterPropertyGroup) {
+      return _makeRow(
+        property.label,
+        Wrap(
+          children: property.members.map<Widget>(_makeLoneCheckbox).toList(),
+        ),
+      );
+    }
+    throw 'unrecognized FilterProperty: $property';
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      children: <Widget>[
+        if (widget.onClose != null)
+          Positioned(
+            child: FlatButton(
+              child: const Icon(Icons.close),
+              onPressed: widget.onClose,
+            ),
+          ),
+        Table(
+          defaultVerticalAlignment: TableCellVerticalAlignment.middle,
+          columnWidths: const <int, TableColumnWidth>{
+            0: IntrinsicColumnWidth(),
+            1: FixedColumnWidth(300.0),
+          },
+          children: widget.propertySource.sheetLayout.map(_makeTableRow).toList(),
+        ),
+      ],
+    );
+  }
```

Comment:
I would like to see a test that toggles one of the filters, and checks that it has been applied. It doesn't need to be hooked up to a TaskGrid, but it should verify that this widget can update the filter datastructure.

Reply:
I added a whole file full of tests for `TaskGridFilter`. I didn't apply it to the grid, but I did isolated applications against mock `CommitStatus` and `QualifiedTask` objects.